### PR TITLE
Go back to using curl to download snapshots

### LINKED ISSUE #1088: Cargo cann't be built in msys2.
Cargo cann't download snapshot in msys2 and fails to build:

```
$ make
/mingw64/bin/python src/etc/dl-snapshot.py x86_64-pc-windows-gnu
Traceback (most recent call last):
  File "src/etc/dl-snapshot.py", line 80, in <module>
    with urlopen(url) as in_file:
  File "src/etc/dl-snapshot.py", line 14, in <lambda>
    urlopen = lambda url: closing(urlopen2(url))
  File "E:/msys64/mingw64/lib/python2.7/urllib2.py", line 154, in urlopen
    return opener.open(url, data, timeout)
  File "E:/msys64/mingw64/lib/python2.7/urllib2.py", line 431, in open
    response = self._open(req, data)
  File "E:/msys64/mingw64/lib/python2.7/urllib2.py", line 454, in _open
    'unknown_open', req)
  File "E:/msys64/mingw64/lib/python2.7/urllib2.py", line 409, in _call_chain
    result = func(*args)
  File "E:/msys64/mingw64/lib/python2.7/urllib2.py", line 1265, in unknown_open
    raise URLError('unknown url type: %s' % type)
urllib2.URLError: <urlopen error unknown url type: https>
Makefile:81: recipe for target 'target/snapshot/bin/cargo.exe' failed
make: *** [target/snapshot/bin/cargo.exe] Error 1
```

### LINKED ISSUE #1346: urllib2 error when installing
I'm following the instructions in the README to install Cargo.

```sh
git clone https://github.com/rust-lang/cargo
cd cargo
git submodule update --init
./.travis.install.deps.sh
./configure --local-rust-root="$PWD"/rustc
make
make install
```

But on the `make` step, it immediately fails with the following Python error. https://gist.github.com/e1859a1b726646296c2a. (Note that I'm running on OS X 10.10.2 with Python 2.7.9.)

See https://github.com/mtschirs/quizduellapi/issues/2#issuecomment-67207159 for possible cause.
> It seems that Python < 2.7.9 didn't validate certificates in the HTTP client.

### PR DESCRIPTION
**Pull request by alexcrichton**
We've had too many problems using urllib to make it worth it.

Closes #1346
Closes #1088
cc #1319 (using a proxy through curl)

### CODE REVIEW

File: src/etc/dl-snapshot.py [hunk 1]
```diff
@@ -77,22 +70,22 @@
...
-    with open(dl_path, 'wb') as out_file:
-        out_file.write(data)
+ret = subprocess.call(["curl", "-o", dl_path, url])
+if ret != 0:
+    raise Exception("failed to fetch url")
+h = hashlib.sha1(open(dl_path, 'rb').read()).hexdigest()
+if h != hash:
+    raise Exception("failed to verify the checksum of the snapshot")
 
-with tarfile.open(dl_path) as tar:
```
Comment (huonw) [inline src/etc/dl-snapshot.py]: Why this change?

**Comment by alexcrichton** [comment]
@huonw ah good catch, updated to keep the relevant changes of  aceba88.

**Comment by retep998** [comment]
This change back to curl has broken the downloading of snapshots for me.
```
$ make
/mingw64/bin/python src/etc/dl-snapshot.py x86_64-pc-windows-gnu
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0curl: (60) SSL certificate problem: unable to get local issuer certificate
More details here: http://curl.haxx.se/docs/sslcerts.html

curl performs SSL certificate verification by default, using a "bundle"
 of Certificate Authority (CA) public keys (CA certs). If the default
 bundle file isn't adequate, you can specify an alternate file
 using the --cacert option.
If this HTTPS server uses a certificate signed by a CA represented in
 the bundle, the certificate verification probably failed due to a
 problem with the certificate (it might be expired, or the name might
 not match the domain name in the URL).
If you'd like to turn off curl's verification of the certificate, use
 the -k (or --insecure) option.
Traceback (most recent call last):
  File "src/etc/dl-snapshot.py", line 75, in <module>
    raise Exception("failed to fetch url")
Exception: failed to fetch url
Makefile:90: recipe for target 'target/snapshot/bin/cargo.exe' failed
make: *** [target/snapshot/bin/cargo.exe] Error 1
```

**Comment by alexcrichton** [comment]
Hm, do you have `curl` install specially? I'm curious why it doesn't have access to the same certificates that other msys2 installations do?

**Comment by bombless** [comment]
Hmmm same issue as @retep998  
I thought it was something wrong in my machine or something.

How did you fix it, @retep998  ?

FWIW, setting proxy on curl doesn't work for my MSYS2.
Not sure if it is about MSYS2 pacman tool/repository.

**Comment by retep998** [comment]
I just manually edited the scripts to pass the -k option to curl. The problem is that with msys2, the msys2 curl works fine, but the mingw64 curl is broken, and to build cargo you need cmake and cmake depends on the mingw64 curl. You could also install an external cmake so you don't need the mingw64 curl installed.

**Comment by bombless** [comment]
Thanks for the info, @retep998 !
Maybe we could use python urllib2 for Windows, and use curl for other systems.
I'll try make that happen by modifying this python script.
